### ISSUE TITLE
wrong inputs passed to run:

### ISSUE THREAD

**Issue by pd3** [snakemake/snakemake#1391]
Using the snakemake version 6.15.1 I encounter what appears to be a rather peculiar bug:

In certain situation when checkpoint is referenced twice, the first occurrence passes incorrect `input` values to a `run` rule. Curiously, correct values are passed to a `shell` rule. A small reproducible test case is given below.

The bug can be reproduced by running
```
rm -rf xxx && mkdir -p xxx && echo -e "1\n2\n3" > xxx/input.txt
snakemake -j1 -s test-bug.txt -d xxx/ -p --nocolor
```
After this, the content of these two new files should be same, but is not (the first is incorrect):
```
$ cat xxx/list.txt
jobs.txt

$ cat xxx/list2.txt
dir/1/ccc.txt
dir/2/ccc.txt
dir/3/ccc.txt
``` 
See the small test pipeline here [test-bug.txt](https://github.com/snakemake/snakemake/files/8041476/test-bug.txt)

Here note I want to use `run` because the `input` array is expected to be very big for my data and the shell line length limit would be exceeded.

**Comment by pd3** [comment]
@johanneskoester I looked in the code but it's too modular and complex, couldn't identify where the `run` vs `shell` inputs get filled. I found out that `update_checkpoint_outputs()` in snakemake/dag.py gets called more times with `run` than with `shell`, but couldn't make anything of that.

Somewhat encouraging is that the `-p` option sees in both cases the correct input dependency
```
rule many_to_one:
    input: dir/1/bbb.txt, dir/2/bbb.txt, dir/3/bbb.txt
    output: list.txt
```
although `run` ends up using `jobs.txt` instead of `dir/1/bbb.txt, dir/2/bbb.txt, dir/3/bbb.txt`.

Since both see the input correctly and `shell` is even able to execute it correctly, this should not be an unsurmountable design problem?

**Comment by youngchanpark** [comment]
Hi again,
I'm getting the problem in the `shell` directive (6.15.5).

**Comment by youngchanpark** [comment]
#855, #1021, and a number of other issues seem to be reporting the same problem. 

This bug seems to have existed for a long time, but still no fixes?

The checkpoint feature is surely one very critical feature for a lot of workflows. Once I found the feature, I changed a lot of the code to utilise it. It seemed like it was working great while developing, but I'm experiencing this problem near the completion of my pipeline, which makes it really inconvenient :/.

I'd advise others who read this to use the checkpoint feature sparingly until the bug is fixed.

**Comment by youngchanpark** [comment]
Hey @pd3,

I ran your minimal example with both 6.15.1 and 7.1.1.
As you said, the output of `xxx/list.txt` only contains `jobs.txt` when the pipeline ran with 6.15.1. But the seems okay in 7.1.1, with the below content.

```
dir/1/bbb.txt
dir/2/bbb.txt
dir/3/bbb.txt
```
This is what you expected right?
From going over the changelog between 6.15.1 and 7.1.1, I can't find which change might've fixed your problem but it looks to be fixed.

I'm not sure whether the update would've fixed my problem, but I've changed the quite code a bit so not sure whether I'd be able to reproduce.
